File: Baekjoon/Swift/DFS-BFS/Q4179.swift
//
//  Q4179.swift
//  BOJ_PS
//
//  Created by 김태훈 on 2021/05/30.
//  불! / BFS

typealias Coord = (r: Int, c: Int)

let nm = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = nm[0]
let m = nm[1]

var maze = [[String]]()
var jq = [Coord]()
var fq = [Coord]()
let d: [Coord] = [(-1,0),(0,-1),(1,0),(0,1)]

func spreadFire(_ next: inout [Coord]) {
    for i in 0 ..< fq.count {
        let now = fq[i]
        
        for j in 0 ..< 4 {
            let nr = now.r + d[j].r
            let nc = now.c + d[j].c
            
            if 0 <= nr && nr < n && 0 <= nc && nc < m {
                if maze[nr][nc] == "." {
                    maze[nr][nc] = "F"
                    next.append((nr,nc))
                }
            }
        }
    }
}

func moveJ(_ next: inout [Coord]) -> Bool {
    for i in 0 ..< jq.count {
        let now = jq[i]
        
        for j in 0 ..< 4 {
            let nr = now.r + d[j].r
            let nc = now.c + d[j].c
            
            if 0 <= nr && nr < n && 0 <= nc && nc < m {
                if maze[nr][nc] == "." {
                    maze[nr][nc] = "J"
                    next.append((nr,nc))
                }
            } else {
                return true
            }
        }
    }
    return false
}

func soluiton() {
    for i in 0 ..< n {
        maze.append(Array(readLine()!).map { String($0) })
        for j in 0 ..< m {
            if maze[i][j] == "J" {
                jq.append((i,j))
            } else if maze[i][j] == "F" {
                fq.append((i,j))
            }
        }
    }
    
    var res = 0
    while true {
        var nextJ = [Coord]()
        var nextF = [Coord]()

        if jq.isEmpty {
            print("IMPOSSIBLE")
            break
        }
        res += 1
        spreadFire(&nextF)
        if moveJ(&nextJ) == true {
            print(res)
            break
        }
        
        jq = nextJ
        fq = nextF
    }
}

soluiton()

File: Baekjoon/Python/BOJ_PS/DFS-BFS/Q4179.py
# 불! / BFS
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
maze = [list(map(str, input().strip())) for _ in range(n)]
d = [(0,-1), (0,1), (-1,0), (1,0)]
jq = []
fq = []

def spreadFire():
    for _ in range(len(fq)):
        r, c = fq.pop(0)
        
        for i in range(4):
            nr = r + d[i][0]
            nc = c + d[i][1]
            
            if 0 <= nr < n and 0 <= nc < m: 
                if maze[nr][nc] == ".":
                    maze[nr][nc] = "F"
                    fq.append((nr,nc))
                
def moveJ():
    for _ in range(len(jq)):
        r, c = jq.pop(0)
        
        for i in range(4):
            nr = r + d[i][0]
            nc = c + d[i][1]
            
            if 0 <= nr < n and 0 <= nc < m: 
                if maze[nr][nc] == ".":
                    maze[nr][nc] = "J"
                    jq.append((nr,nc)) 
            else:
                return True
    return False

def solution():
    for i in range(n):
        for j in range(m):
            if maze[i][j] == 'J':
                jq.append((i,j))
            elif maze[i][j] == 'F':
                fq.append((i,j))

    res = 0
    while True:
        res += 1
        if len(jq) == 0:
            print('IMPOSSIBLE')
            break
        spreadFire()
        if moveJ() == True:
            print(res)
            break
solution()